core/src/main/java/org/elasticsearch/gateway/Gateway.java

```diff
@@ -66,9 +77,9 @@ public Gateway(Settings settings, ClusterService clusterService, NodeEnvironment
     }
 
     public void performStateRecovery(final GatewayStateRecoveredListener listener) throws GatewayException {
-        ObjectHashSet<String> nodesIds = new ObjectHashSet<>(clusterService.state().nodes().masterNodes().keys());
-        logger.trace("performing state recovery from {}", nodesIds);
-        TransportNodesListGatewayMetaState.NodesGatewayMetaState nodesState = listGatewayMetaState.list(nodesIds.toArray(String.class), null).actionGet();
+        String[] nodesIds = new ObjectHashSet<>(clusterService.state().nodes().masterNodes().keys()).toArray(String.class);
```

Comment:
we are not going to have duplicates in map keys? Why not directly `clusterService.state().nodes().masterNodes().keys().toArray()`?